vm/devices/storage/disk_nvme/nvme_driver/src/driver.rs

```diff
@@ -408,17 +465,24 @@ impl<T: DeviceBacking> NvmeDriver<T> {
     }
 
     /// Gets the namespace with namespace ID `nsid`.
-    pub async fn namespace(&self, nsid: u32) -> Result<Namespace, NamespaceError> {
-        Namespace::new(
-            &self.driver,
-            self.admin.as_ref().unwrap().clone(),
-            self.rescan_event.clone(),
-            self.identify.clone().unwrap(),
-            &self.io_issuers,
-            &self.device_id,
-            nsid,
-        )
-        .await
+    pub async fn namespace(&mut self, nsid: u32) -> Result<Arc<Namespace>, NamespaceError> {
+        // Even if namespace might have been previously restored,
+        // query Identify again to prevent possible mismatch.
+        // TODO: Check if we will process Namespace Change AEN instead.
+        let ns_new = Arc::new(
+            Namespace::new(
+                &self.driver,
+                self.admin.as_ref().unwrap().clone(),
+                self.rescan_event.clone(),
+                self.identify.clone().unwrap(),
+                &self.io_issuers,
+                nsid,
+            )
+            .await?,
+        );
+
+        self.namespaces.push(ns_new.clone());
```

Comment:
No, please don't keep the namespaces around. I believe this will leak memory when namespaces are removed by the host.

Reply:
That's why I put Arc here - one reference is discarded when servicing state is destroyed, another should be dropped when object is replaced with the new namespace?

Let me check if I missed something or if it can be redone.

Reply:
I don't know why you insist on trying to keep parts of  this disabled namespace save/restore around. The design is broken, and you should just rip it out. You can add it back in a subsequent change once the design looks good.

Reply:
Done